rust/src/dumpfile.rs

```diff
@@ -0,0 +1,391 @@
+        };
+        Ok(Entry {
+            path,
+            uid,
+            gid,
+            mode,
+            mtime,
+            item,
+            xattrs,
+        })
+    }
+
+    /// Remove internal entries
+    /// FIXME: This is arguably a composefs-info dump bug?
+    pub fn filter_special(mut self) -> Self {
+        self.xattrs.retain(|v| match (v.key.as_bytes(), &*v.value) {
+            (b"trusted.overlay.opaque" | b"user.overlay.opaque", b"x") => false,
+            _ => true,
+        });
+        self
+    }
+}
+
+impl<'p> Item<'p> {
+    pub(crate) fn size(&self) -> u64 {
+        match self {
+            Item::Regular { size, .. } | Item::Directory { size, .. } => *size,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn nlink(&self) -> u32 {
+        match self {
+            Item::Regular { nlink, .. } => *nlink,
+            Item::Device { nlink, .. } => *nlink,
+            Item::Symlink { nlink, .. } => *nlink,
+            Item::Directory { nlink, .. } => *nlink,
+            Item::Fifo { nlink, .. } => *nlink,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn rdev(&self) -> u32 {
+        match self {
+            Item::Device { rdev, .. } => *rdev,
+            _ => 0,
+        }
+    }
+
+    pub(crate) fn payload(&self) -> Option<&Path> {
+        match self {
+            Item::Symlink { target, .. } => Some(target),
+            Item::Hardlink { target } => Some(target),
+            _ => None,
+        }
+    }
+}
+
+impl Display for Mtime {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}", self.sec, self.nsec)
+    }
+}
+
+impl<'p> Display for Entry<'p> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        escape(f, self.path.as_os_str().as_bytes())?;
+        write!(
+            f,
+            " {} {:o} {} {} {} {} {} ",
+            self.item.size(),
+            self.mode,
+            self.item.nlink(),
+            self.uid,
+            self.gid,
+            self.item.rdev(),
+            self.mtime,
+        )?;
+        if let Some(payload) = self.item.payload() {
+            write!(f, "{payload:?} ")?;
+        } else {
+            write!(f, "- ")?;
+        }
+        match &self.item {
+            Item::Regular {
+                fsverity_digest,
+                inline_content,
+                ..
+            } => {
+                if let Some(content) = inline_content {
+                    escape(f, content)?;
+                    f.write_char(' ')?;
+                } else {
+                    write!(f, "- ")?;
+                }
+                let fsverity_digest = fsverity_digest.as_deref().unwrap_or("-");
+                write!(f, "{fsverity_digest}")?;
+            }
+            _ => {
+                write!(f, "- -")?;
+            }
+        }
+        for xattr in self.xattrs.iter() {
+            f.write_char(' ')?;
+            escape(f, xattr.key.as_bytes())?;
```

Comment:
This should also escape `=` in the key, so the split is guaranteed to work.

Reply:
Good catch (here and elsewhere!) - fixed.